concurrency/WatchableFactory.java

```diff
@@ -0,0 +1,86 @@
+package concurrency;
+
+import com.a9.cpx.monitoring.pmet.StatusChangeHandler;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public abstract class WatchableFactory<T extends Watchable> {
+
+    List<WatchableRegistry> watchableRegistry;
+    Map<Attributes,T> watchableMap = new ConcurrentHashMap<Attributes, T>();
+    StatusChangeHandler statusChangeHandler;
+
+    public void init() {
+    }
+
+    public List<WatchableRegistry> getWatchableRegistry() {
+        return watchableRegistry;
+    }
+
+    public void setWatchableRegistry(List<WatchableRegistry> watchableRegistry) {
+        this.watchableRegistry = watchableRegistry;
+    }
+
+    public StatusChangeHandler getStatusChangeHandler() {
+        return statusChangeHandler;
+    }
+
+    public void setStatusChangeHandler(StatusChangeHandler statusChangeHandler) {
+        this.statusChangeHandler = statusChangeHandler;
+    }
+
+    protected void register(Watchable watchable) {
+        if(watchableRegistry != null) {
+            for(WatchableRegistry w : watchableRegistry) {
+                w.register(watchable);
+            }
+        }
+
+        if(statusChangeHandler != null) {
+            watchable.setStatusChangeHandler(statusChangeHandler);
+        }
+    }
+
+    protected boolean unregister(Watchable watchable)  {
+        boolean ret = true;
+        if(watchableRegistry != null) {
+            for(WatchableRegistry w : watchableRegistry) {
+                ret&= w.unregister(watchable);
+            }
+        }
+        
+        return ret;
+    }
+
+    public abstract T newInstance(Attributes attribues);
+
+    public T getWatchable(Attributes attributes) {
+        T watchable = watchableMap.get(attributes);
+        if(watchable == null) {
+            synchronized(this){
+                watchable = watchableMap.get(attributes);
+                if(watchable == null){
+                    watchable = newInstance(attributes);
+                    watchableMap.put(attributes, watchable);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 These lines use `ConcurrentHashMap`&#39;s **sequential** API and are protected by `synchronized` to ensure they execute together: **64 and 67**. This synchronization may reduce performance by limiting parallelism and does not take advantage of `ConcurrentHashMap`&#39;s **concurrent** API.

**Fix**  
 Consider replacing the check and `put()` with `putIfAbsent()` and then removing the synchronization. The check is at line: **65**.

**More info**  
 `putIfAbsent()` is thread-safe and puts the value only if the `ConcurrentHashMap` does not contain the key. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/HistoryFileManager.java#L927-L932) (external link).